### PR TITLE
vcr: use go-vcr package to setup vcr recorder

### PR DESCRIPTION
**Pull request by gemmahou**
### Change description

In this PR, I implemented the same thing as in #1345, with a different approach. I incorporated [go-vcr ](https://github.com/dnaeon/go-vcr) package and used its recorder.

### Tests you have done

- [x] Run `make ready-pr` to ensure this PR is ready for review.
- [ ] Perform necessary E2E testing for changed resources.

### CODE REVIEW

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
...n.publicerrors.PublicErrorProtoUtils.newErrorBuilder(PublicErrorProtoUtils.java:2158)\n\tat com.google.cloud.control.common.publicerrors.PublicErrorProtoUtils.newErrorBuilder(PublicErrorProtoUtils.java:2150)\n\tat com.google.cloud.control.common.publicerrors.PublicErrorProtoUtils.createResourceNotFoundError(PublicErrorProtoUtils.java:194)\n\tat com.google.cloud.control.entities.clhbridge.ClhBridgeEntityLoader.lambda$getEntityByReference$0(ClhBridgeEntityLoader.java:164)\n\tat java.base/java.util.Optional.orElseThrow(Unknown Source)\n\tat com.google.cloud.control.entities.clhbridge.ClhBridgeEntityLoader.getEntityByReference(ClhBridgeEntityLoader.java:161)\n\tat com.google.cloud.control.services.clhbridge.ClhBridgeGetResourceAction$Handler.runAttempt(ClhBridgeGetResourceAction.java:390)\n\tat com.google.cloud.control.services.clhbridge.ClhBridgeGetResourceAction$Handler.runAttempt(ClhBridgeGetResourceAction.java:267)\n\tat com.google.cloud.cluster.metastore.RetryingMetastoreTransactionExecutor$1.runAttempt(RetryingMetastoreTransactionExecutor.java:94)\n\tat com.google.cloud.cluster.metastore.MetastoreRetryLoop.runHandler(MetastoreRetryLoop.java:523)\n\t...Stack trace is shortened.\n"
```
Comment (justinsb) [inline vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:56]: Any way to remove these lines (without too much trouble?)

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml [hunk 2]
```diff
@@ -0,0 +1,543 @@
...
+        proto_minor: 0
+        transfer_encoding: []
+        trailer: {}
+        content_length: -1
+        uncompressed: true
+        body: |
+            {
+              "error": {
+                "code": 404,
+                "message": "The resource 'projects/cnrm-yuhou/regions/us-central1/nodeTemplates/computenodetemplate-le4q3bkhysnhydb2zckq' was not found",
```
Comment (justinsb) [inline vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:50]: Can we normalize the project name here to e.g. `example-project` ?

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml [hunk 3]
```diff
@@ -0,0 +1,543 @@
...
+        proto_major: 1
+        proto_minor: 1
+        content_length: 417
+        transfer_encoding: []
+        trailer: {}
+        host: compute.googleapis.com
+        remote_addr: ""
+        request_uri: ""
+        body: |
+            {"cpuOvercommitType":"NONE","description":"Node template for sole tenant nodes running in us-central1, with 96vCPUs and any amount of memory on any machine type.","name":"computenodetemplate-le4q3bkhysnhydb2zckq","nodeAffinityLabels":{"cnrm-test":"true","managed-by-cnrm":"true","memory_guarantee":"false"},"nodeTypeFlexibility":{"cpus":"96","memory":"any"},"region":"projects/cnrm-yuhou/global/regions/us-central1"}
```
Comment (justinsb) [inline vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:95]: Maybe replace `/${projectId}/` with `/example-project-id/` would work?

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml [hunk 4]
```diff
@@ -0,0 +1,543 @@
...
+        body: |
+            {
+              "kind": "compute#operation",
+              "id": "4057836747365686959",
+              "name": "operation-1710202944238-6136ba6fc31ca-90ae9715-222761eb",
+              "operationType": "compute.nodeTemplates.insert",
+              "targetLink": "https://www.googleapis.com/compute/beta/projects/cnrm-yuhou/regions/us-central1/nodeTemplates/computenodetemplate-le4q3bkhysnhydb2zckq",
+              "targetId": "538944551121985199",
+              "status": "RUNNING",
+              "user": "yuhou@google.com",
```
Comment (justinsb) [inline vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:121]: Nit: should probably replace your username also

**Comment by justinsb** [comment]
A few nit but LGTM... can we add a VCR test as a GitHub actions presubmit (maybe just running one subtest to start with?)

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml (hunk 1 above)
Comment (gemmahou) [reply vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:56]: Addressed in PR #1349

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml (hunk 2 above)
Comment (gemmahou) [reply vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:50]: Will be address in PR #1349

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml (hunk 3 above)
Comment (gemmahou) [reply vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:95]: Will be address in PR #1349

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml (hunk 4 above)
Comment (gemmahou) [reply vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml:121]: Addressed in PR #1349

File: config/tests/samples/create/harness.go [hunk 5]
```diff
@@ -75,6 +77,8 @@ type Harness struct {
 	// gcpAccessToken is set to the oauth2 token to use for GCP, primarily when GCP is mocked.
 	gcpAccessToken string
 	kccConfig      kccmanager.Config
+
+	Rec *recorder.Recorder
```
[the remark this answers is not in the document]
Comment (gemmahou) [reply config/tests/samples/create/harness.go]: Updated!

File: config/tests/samples/create/harness.go [hunk 6]
```diff
@@ -306,17 +312,37 @@ func NewHarness(ctx context.Context, t *testing.T) *Harness {
...
+				ret = &http.Client{Transport: t.(http.RoundTripper)}
+			}
+			dir := "vcr/fixtures/cassette/"
+			testName := strings.ReplaceAll(t.Name(), "/", "_")
+			opts := &recorder.Options{
+				CassetteName:  dir + testName,
+				Mode:          recorder.ModeRecordOnly,
+				RealTransport: ret.Transport,
+			}
+			r, _ := recorder.NewWithOptions(opts)
```
[the remark this answers is not in the document]
Comment (gemmahou) [reply config/tests/samples/create/harness.go]: Updated to handle the error

File: tests/e2e/unified_test.go [hunk 7]
```diff
@@ -120,6 +120,16 @@ func TestAllInSeries(t *testing.T) {
 				h := create.NewHarness(ctx, t)
 				project := h.Project
 
+				if os.Getenv("E2E_GCP_TARGET") == "vcr" {
+					// Stop recording after tests finish and write to cassette
+					defer func() {
```
[the remark this answers is not in the document]
Comment (gemmahou) [reply tests/e2e/unified_test.go]: `defer` is suggested in the go-vcr readme. But I don't see any difference by using t.Cleanup. I updated here for now.

File: vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+---
+version: 2
+interactions:
```
[the remark this answers is not in the document]
Comment (gemmahou) [reply vcr/fixtures/cassette/TestAllInSeries_fixtures_computenodetemplate.yaml]: The file is generated by go-vcr default recorder, and I think the format matters when we use the replaying mode to parse request/response from cassette. I prefer leave it for now. We can make that as an improvement in the future.

**Comment by gemmahou** [comment]
> A few nit but LGTM... can we add a VCR test as a GitHub actions presubmit (maybe just running one subtest to start with?)

I'm thinking about that too, I'll have that setup in a future PR and see how it works.  I'm also thinking about if we can migrate VCR test into our prow job so that we can save cassette files in GCS private storage, similar to artifacts [http logs](https://pantheon.corp.google.com/storage/browser/cnrm-prow/logs/periodic-cnrm-test-probes/1768200554405695488/artifacts/http-logs?pageState=(%22StorageObjectListTable%22:(%22f%22:%22%255B%255D%22))&mods=pan_ng2&prefix=&forceOnObjectsSortingFiltering=false).

**Pull request merged**